EIPS/eip-4788.md

```diff
@@ -52,37 +53,63 @@ Validity is guaranteed from the consensus layer, much like how withdrawals are h
 At the start of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. before processing any transactions),
 write the parent beacon root provided in the block header into the storage of the contract at `HISTORY_STORAGE_ADDRESS`.
 
-The root itself is used as a key into the contract's storage and the timestamp of the header is written as the key's value.
-The timestamp (a 64-bit unsigned integer value) is encoded as 32 bytes in big-endian format.
+In order to bound the storage used by this precompile, two ring buffers are used: one to track the latest root at a given index and another to track
+the latest timestamp at a given index.
+
+To derive the index `root_index` into the root ring buffer, the timestamp (a 64-bit unsigned integer value) is reduced modulo `HISTORICAL_ROOTS_LENGTH`.
+To derive the index `timestamp_index` into the timestamp ring buffer, add `HISTORICAL_ROOTS_LENGTH` to the index into the root ring buffer.
+Both resulting 64-bit unsigned integers should be encoded as 32 bytes in big-endian format when writing to the storage.
+
+The 32 bytes of the `parent_beacon_block_root` (as provided) are the value to write behind the `root_index`.
+The timestamp from the header, encoded as 32 bytes in big-endian format, is the value to write behind the `timestamp_index`.
 
 In Python pseudocode:
 
 ```python
+timestamp_reduced = block_header.timestamp % HISTORICAL_ROOTS_LENGTH
+timestamp_extended = timestamp_reduced + HISTORICAL_ROOTS_LENGTH
+root_index = to_uint256_be(timestamp_reduced)
+timestamp_index = to_uint256_be(timestamp_extended)
+
 parent_beacon_block_root = block_header.parent_beacon_block_root
-timestamp = to_uint256_be(block_header.timestamp)
+timestamp_as_uint256 = to_uint256_be(block_header.timestamp)
 
-sstore(HISTORY_STORAGE_ADDRESS, parent_beacon_block_root, timestamp)
+sstore(HISTORY_STORAGE_ADDRESS, root_index, parent_beacon_block_root)
+sstore(HISTORY_STORAGE_ADDRESS, timestamp_index, timestamp_as_uint256)
 ```
 
 #### New stateful precompile
 
-Beginning at the execution timestamp `FORK_TIMESTAMP`, the code and storage at `HISTORY_STORAGE_ADDRESS` constitute a "stateful" precompile.
+Beginning at the execution timestamp `FORK_TIMESTAMP`, a "stateful" precompile is deployed at `HISTORY_STORAGE_ADDRESS`.
+
+Callers of the precompile should provide the `timestamp` they are querying encoded as 32 bytes in big-endian format.
 
-Callers of the precompile should provide the `root` they are querying encoded as 32 bytes.
+Given this input, the precompile reduces the `timestamp` in the same way during the write routine and first checks if
```

Comment:
Given the timestamp is checked before the root is read, it seems a bit more natural to have it ordered timestamp buffer followed by the root buffer. But just a stylistic suggestion

i.e.
1. Check something
2. if check passes,  move on the next thing to read

Reply:
sure I can update